File: Coding-Interviews/26-树的子结构.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    bool match(TreeNode* A, TreeNode* B)
    {
        if (!B)
            return true;
        if (!A)
            return false;
        return A->val == B->val && match(A->left, B->left) && match(A->right, B->right);
    }
    bool isSubStructure(TreeNode* A, TreeNode* B)
    {
        if (!A && !B)
            return true;
        if (!A || !B)
            return false;
        return match(A, B) || isSubStructure(A->left, B) || isSubStructure(A->right, B);
    }
};

File: Coding-Interviews/26-树的子结构.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        def match(A: TreeNode, B: TreeNode):
            if not B:
                return True
            if not A:
                return False
            if A.val == B.val:
                return match(A.left, B.left) and match(A.right, B.right)
        if not B:
            return False
        if not A:
            return False
        return match(A, B) or self.isSubStructure(A.left, B) or self.isSubStructure(A.right, B)
